src/wallet/rpcwallet.cpp

```diff
@@ -598,6 +652,102 @@ UniValue listaddresses(const UniValue& params, bool fHelp)
+                // double-check that the ufvkid we get from address metadata is actually
+                // associated with the mnemonic HD chain
+                auto ufvkCheck = pwalletMain->mapUnifiedAccountKeys.find(
+                    std::make_pair(hdChain.value().GetSeedFingerprint(), account.value())
+                );
+                if (ufvkCheck != pwalletMain->mapUnifiedAccountKeys.end() && ufvkCheck->second == ufvkid) {
+                    UniValue unified_group(UniValue::VOBJ);
+                    unified_group.pushKV("account", uint64_t(account.value()));
+
+                    UniValue unified_addrs(UniValue::VARR);
+                    auto ufvk = pwalletMain->GetUnifiedFullViewingKey(ufvkid).value();
+                    for (const auto& [j, receiverTypes] : addrmeta.GetKnownReceiverSetsByDiversifierIndex()) {
+                        // We know we can use std::get here safely because we previously
+                        // generated a valid address for this diversifier & set of
+                        // receiver types.
+                        UniValue addrEntry(UniValue::VOBJ);
+                        auto addr = std::get<std::pair<libzcash::UnifiedAddress, diversifier_index_t>>(
+                            ufvk.Address(j, receiverTypes)
+                        );
+                        UniValue receiverTypesEntry(UniValue::VARR);
+                        for (auto t : receiverTypes) {
+                            switch(t) {
+                                case ReceiverType::P2PKH:
+                                    receiverTypesEntry.push_back("p2pkh");
+                                    break;
+                                case ReceiverType::P2SH:
+                                    receiverTypesEntry.push_back("p2sh");
+                                    break;
+                                case ReceiverType::Sapling:
+                                    receiverTypesEntry.push_back("sapling");
+                                    break;
+                                case ReceiverType::Orchard:
+                                    receiverTypesEntry.push_back("orchard");
+                                    break;
+                            }
+                        }
+                        {
+                            UniValue jVal;
+                            jVal.setNumStr(ArbitraryIntStr(std::vector(j.begin(), j.end())));
+                            addrEntry.pushKV("diversifier_index", jVal);
+                        }
+                        addrEntry.pushKV("receiver_types", receiverTypesEntry);
+                        addrEntry.pushKV("address", keyIO.EncodePaymentAddress(addr.first));
+                        unified_addrs.push_back(addrEntry);
+                    }
+                    unified_group.pushKV("addresses", unified_addrs);
+                    unified_groups.push_back(unified_group);
+                }
```

Comment:
This is okay for now, but I think that perhaps we should have a section for each account even if that section is empty? It's useful information for a user to know if they need to generate a new account, or simply to generate an address for an account that they've already created, if that account doesn't yet have any addresses.

Reply:
+1; this can happen after 4.7.0, so non-blocking.